pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -544,33 +545,64 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 
 	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
 	updateMin := 0
+	maxUnavailable := 1
 	if set.Spec.UpdateStrategy.RollingUpdate != nil {
 		updateMin = int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+		maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(replicaCount), false)
+		if err != nil {
+			return &status, err
+		}
+		// if the feature was enabled and then later disabled, MaxUnavailable may have a value
+		// other than 1. Reset maxUnavailable to 1 to revert to the expected behavior when this
+		// feature gate is not enabled.
+		if !utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
+			maxUnavailable = 1
+		}
 	}
-	// we terminate the Pod with the largest ordinal that does not match the update revision.
+
+	// Collect all targets in the range between the partition and Spec.Replicas. Count any targets in that range that are unhealthy
+	// i.e. terminated or not running and ready as unavailable). Select the (MaxUnavailable - Unavailable) Pods, in order with respect
+	// to their ordinal for termination. Delete those pods and count the successful deletions. Update the status with the correct
+	// number of deletions.
+	unavailablePods := 0
 	for target := len(replicas) - 1; target >= updateMin; target-- {
+		if !isHealthy(replicas[target]) {
+			unavailablePods++
+		}
+	}
 
-		// delete the Pod if it is not already terminating and does not match the update revision.
-		if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
+	// Now we need to delete MaxUnavailable- unavailablePods
+	// start deleting one by one starting from the highest ordinal first
+	podsToDelete := maxUnavailable - unavailablePods
+	deletedPods := 0
+	for target := len(replicas) - 1; target >= updateMin; target-- {
+
+		// delete the Pod if it is healthy and the revision doesnt match the target
+		if getPodRevision(replicas[target]) != updateRevision.Name && isHealthy(replicas[target]) {
 			klog.V(2).Infof("StatefulSet %s/%s terminating Pod %s for update",
 				set.Namespace,
 				set.Name,
 				replicas[target].Name)
-			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[target]); err != nil {
+				return nil, err
+			}
+			deletedPods++
+
 			status.CurrentReplicas--
-			return &status, err
 		}
 
-		// wait for unhealthy Pods on update
-		if !isHealthy(replicas[target]) {
+		// If at anytime, total number of unavailable Pods exceeds maxUnavailable,
+		// we stop deleting more Pods for Update
+		if deletedPods == podsToDelete {
 			klog.V(4).Infof(
```

Comment:
since you are already changing the log line: use structured logging `klog.InfoS` https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions-to-structured-equivalent